### PR TITLE
DAOS-6802 build: Use a dedicated dockerfile for checkpatch.

### PR DESCRIPTION
**Pull request by ashleypittman**

### CODE REVIEW

File: utils/docker/Dockerfile.checkpatch
Comment (brianjmurrell) [inline utils/docker/Dockerfile.checkpatch:8]: If you are going to have a dedicated Dockerfile for this, why not use a distro with *much* newer tools, like, say, `fedora:latest`?

File: utils/docker/Dockerfile.checkpatch
Comment (brianjmurrell) [inline utils/docker/Dockerfile.checkpatch:10]: If you move this lower in the file, you will get more re-use at the higher layers.  I.e. if you do install first then cache-bust then upgrade, you will get longer life out of that pre-upgrade layer and less work in the lower layers.

Maybe we want to consider multiple layers of cache-busting.  One weekly and one daily (and/or one at the sub-hour value).  That way you are getting to-the-day (or hour or minute) updates in that final layer, but that operation should go quickly because you are leveraging a weekly layer above it.

File: utils/docker/Dockerfile.checkpatch
Comment (brianjmurrell) [inline utils/docker/Dockerfile.checkpatch:36]: All of this probably comes packaged in Fedora.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: utils/docker/Dockerfile.checkpatch
Comment (ashleypittman) [reply utils/docker/Dockerfile.checkpatch:10]: For checkpatch/codespell weekly is probably frequent enough, and the package list is short enough I didn't see much benefit in doing this multi-layered.  It really is a very small amount of work in this file now.

File: utils/docker/Dockerfile.checkpatch
Comment (ashleypittman) [reply utils/docker/Dockerfile.checkpatch:36]: pygithub is packaged in centos7 already, however I found if was missing a "dismiss" function that we use so I stuck with the pip version.  For codespell I thought pip was probably more up-to-date than centos although this probably doesn't matter all that much.

File: utils/docker/Dockerfile.checkpatch
Comment (ashleypittman) [reply utils/docker/Dockerfile.checkpatch:8]: I tried quite hard with this PR to use centos8, for that very reason, but I couldn't find a way of getting ShellCheck installed, it seems to have moved from yum/dnf to using snap as an installer which is new to me and doesn't seem to work in docker.  Doing a periodic update to latest would benefit here, but having tried I took the safe option and extracted the code/versions we're already using.

Comment (JohnMalmberg) [reply utils/docker/Dockerfile.checkpatch:8]: I am using Shellcheck and even some of the other linters that Daos does not currently have enabled with a Fedora:latest Dockerfile for other projects.

I have not tried the Checkpatch yet.

Comment (brianjmurrell) [reply utils/docker/Dockerfile.checkpatch:8]: > I tried quite hard with this PR to use centos8,

centos8 is going to eventually have the same geriatric issues that centos7 has currently.  The whole "EL" project simply is not updated frequently enough with newer (not just bug-fixed) software as it ages and therefore suitable for this task.

> for that very reason, but I couldn't find a way of getting ShellCheck installed

It will be installed with `fedora:latest`.  I use Fedora as my main platform and I use ShellCheck all the time on it.  It does catch way more than El7's ageing ShellCheck, FWIW.

> it seems to have moved from yum/dnf to using snap as an installer

That wouldn't be the "official" EL8 ShellCheck.  The Snap store version is likely not maintained by RH and in this manner is a "third-party" store.  That said, I am quite pessimistic on both Snap and Flatpak packaging systems, as they are both _bundling_ packaging systems, which the core software you want to install also installs copies of all of it's dependencies.  That then puts a burden and responsibility on those package maintainers to not only update their packages when the core software is updated, but they also have to be paying attention to the upstreams of **all** of the software they are bundling and put out new releases when any of those upstreams release a security update.  I am **not at all** convinced that those packagers are taking on the responsibility sufficiently and actually paying attention to their bundled software upstreams.

https://bugzilla.redhat.com/show_bug.cgi?id=1772274 would be the RH perspective on ShellCheck for EL8.

> which is new to me and doesn't seem to work in docker.

And we don't want to use, anyway, IMO, for the reasons above.

> Doing a periodic update to latest would benefit here, but having tried I took the safe option and extracted the code/versions we're already using.

I would say that if we are going to this effort to have a dedicated linting Dockerfile, let's do it with recent tool versions, and we are only going to be able to get recent tool versions with something like Fedora.

File: utils/docker/Dockerfile.checkpatch
Comment (brianjmurrell) [reply utils/docker/Dockerfile.checkpatch:10]: For this effort that's probably fair enough.  I think I was making my comment more generically though, probably more applicable to `Dockerfile`s that build our software and then perhaps only apply that here in any effort to maintain a single structure for all of our Dockerfiles.

The other thing to consider in the frequency of updates is security and how long you want a given Docker image to be vulnerable to a particular security issue.

File: utils/docker/Dockerfile.checkpatch
Comment (brianjmurrell) [reply utils/docker/Dockerfile.checkpatch:36]: Did you check the version packaged in Fedora?

File: utils/docker/Dockerfile.checkpatch
Comment (ashleypittman) [reply utils/docker/Dockerfile.checkpatch:8]: > > I tried quite hard with this PR to use centos8,
> 
> centos8 is going to eventually have the same geriatric issues that centos7 has currently. The whole "EL" project simply is not updated frequently enough with newer (not just bug-fixed) software as it ages and therefore suitable for this task.
> 
> > for that very reason, but I couldn't find a way of getting ShellCheck installed
> 
> It will be installed with `fedora:latest`. I use Fedora as my main platform and I use ShellCheck all the time on it. It does catch way more than El7's ageing ShellCheck, FWIW.
> 
> > it seems to have moved from yum/dnf to using snap as an installer
> 
> That wouldn't be the "official" EL8 ShellCheck. The Snap store version is likely not maintained by RH and in this manner is a "third-party" store. That said, I am quite pessimistic on both Snap and Flatpak packaging systems, as they are both _bundling_ packaging systems, which the core software you want to install also installs copies of all of it's dependencies. That then puts a burden and responsibility on those package maintainers to not only update their packages when the core software is updated, but they also have to be paying attention to the upstreams of **all** of the software they are bundling and put out new releases when any of those upstreams release a security update. I am **not at all** convinced that those packagers are taking on the responsibility sufficiently and actually paying attention to their bundled software upstreams.
> 
> https://bugzilla.redhat.com/show_bug.cgi?id=1772274 would be the RH perspective on ShellCheck for EL8.
> 
> > which is new to me and doesn't seem to work in docker.
> 
> And we don't want to use, anyway, IMO, for the reasons above.

Agreed, on all counts.  I didn't find that redhat ticket but it does explain why I had problems.

> > Doing a periodic update to latest would benefit here, but having tried I took the safe option and extracted the code/versions we're already using.
> 
> I would say that if we are going to this effort to have a dedicated linting Dockerfile, let's do it with recent tool versions, and we are only going to be able to get recent tool versions with something like Fedora.

That wasn't the problem I set out to solve, but it seems splitting this dockerfile out is a essential first step for us to get there.

File: utils/docker/Dockerfile.checkpatch
Comment (ashleypittman) [reply utils/docker/Dockerfile.checkpatch:10]: That's a good point, in fact it sounds like this is demonstrated quite well with VM images in that they're updated every time but the base image are quite old.  I guess we want one level to update "frequently" to ensure the software is up-to-date and another level to force a complete rebuild "infrequently" to ensure the update doesn't start taking too long.

Comment (brianjmurrell) [reply utils/docker/Dockerfile.checkpatch:10]: > sounds like this is demonstrated quite well with VM images in that they're updated every time but the base image are quite old.

Yeah, we really have to get (and we are getting there -- just so much to do and so little resources to apply) to a point of having completely hands-off updating of the images on the CI test nodes.  Then we can update them at whatever frequency makes sense.  For IML, we used to do it once every 24h.  So one test run in a 24h time window would take a little bit longer than the rest on every cluster as the base was rebuilt.  I never really saw it as being all that advantageous (in respect to the effort to implement it) to make the image updating and replacing asynchronous so that even the 1 test run in the 24 time window didn't incur the extra build time.  But that could be done if it was really seen as an issue.

**Comment by ashleypittman** [comment]
Updated to latest fedora.  It still loads codespell from pip because the options we're using are supported in the fedora version.

I've verified that the python checks work, however not checkpatch itself or shellcheck but I can test this later by adding issues into this ticket.

**Comment by brianjmurrell** [comment]
> Updated to latest fedora. It still loads codespell from pip because the options we're using are supported in the fedora version.

What version are you getting out of pypi?

**Comment by ashleypittman** [comment]
> > Updated to latest fedora. It still loads codespell from pip because the options we're using are supported in the fedora version.
> 
> What version are you getting out of pypi?

This just gets more confusing.

In this PR fedora was getting 1.16.0-1.fc31 which is an older version
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-4648/22/pipeline/32

Trying locally however fedora is giving me 2.0.0-1.fc33 which appears to have the option, and is similar to pip which is codespell-2.0.0

**Comment by ashleypittman** [comment]
This is working correctly now, and checkpatch/shellcheck/pylint and codespell have all been verified to work.

File: utils/docker/Dockerfile.checkpatch
Comment (brianjmurrell) [inline utils/docker/Dockerfile.checkpatch:8]: Didn't we discuss using `:latest` here so that it's not somebody's job to have (to remember) to fix this every 6 months?

File: Jenkinsfile
Comment (brianjmurrell) [inline Jenkinsfile:565]: What about security updates?  Are we going to potentially live with an image that could have a security vulnerability in it for up to a week?

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: utils/docker/Dockerfile.checkpatch
Comment (ashleypittman) [reply utils/docker/Dockerfile.checkpatch:8]: I can do whichever you prefer, but I think we need to add --pull to the Jenkinsfile if we do that.

File: Jenkinsfile
Comment (ashleypittman) [reply Jenkinsfile:565]: Yes, that's what I'm proposing.  How often do we update Jenkins itself?

Comment (brianjmurrell) [reply Jenkinsfile:565]: Weekly, as needed (there isn't a new version out every week I don't think).  But Jenkins is entirely different as it disrupts the entire team to update it.

Comment (ashleypittman) [reply Jenkinsfile:565]: How often do you think would be appropriate and do we have an internal up-to-date fedora:31 mirror?  Going outside the network for every PR seems overkill and unreliable.

How often does our mirror update for the repos that we do store cached versions of?

Comment (brianjmurrell) [reply Jenkinsfile:565]: > How often do you think would be appropriate

For security purposes, probably at least daily.

> and do we have an internal up-to-date fedora:31 mirror?

Surely you meant `fedora:latest` which is `fedora:33` at the moment, and yes, our caching proxy server will always be as up-to-date as the clients that update through it.

> Going outside the network for every PR seems overkill and unreliable.

Yes, we should always be using our internal mirror/caching-proxy-server.  That means having to make our Dockerfiles live both inside of our network with the luxury of our mirror as well as outside using regular upstream mirrors.  There are examples of how we do this in the Dockerfile.centos.7 file.
 
> How often does our mirror update for the repos that we do store cached versions of?

The mirror is constantly being updated as clients update through it.

File: utils/docker/Dockerfile.checkpatch
Comment (brianjmurrell) [inline utils/docker/Dockerfile.checkpatch:33]: https://dnf.readthedocs.io/en/latest/command_ref.html#upgrade-command-label

`update` is a deprecated alias for `upgrade`.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
-1 for the update/upgrade but also just to make sure this does not land before #4673.

File: utils/docker/Dockerfile.checkpatch
Comment (ashleypittman) [reply utils/docker/Dockerfile.checkpatch]: Done.

**Comment by ashleypittman** [comment]
> -1 for the update/upgrade but also just to make sure this does not land before #4673.

That's orthogonal, the damage there is already done.

**Comment by brianjmurrell** [comment]
> > -1 for the update/upgrade but also just to make sure this does not land before #4673.
> 
> That's orthogonal, the damage there is already done.

Only if a full docker build from `fedora:latest` has been done on master.  I can't think of why a master run would have fully rebuilt a docker image starting with `fedora:latest` any time recently.  If one has been done, it would be interesting to investigate why, because the Dockerfile using `fedora:latest` should not have changed to any degree resulting in a complete image rebuild.

**Comment by ashleypittman** [comment]
> > > -1 for the update/upgrade but also just to make sure this does not land before #4673.
> > 
> > 
> > That's orthogonal, the damage there is already done.
> 
> Only if a full docker build from `fedora:latest` has been done on master. I can't think of why a master run would have fully rebuilt a docker image starting with `fedora:latest` any time recently. If one has been done, it would be interesting to investigate why, because the Dockerfile using `fedora:latest` should not have changed to any degree resulting in a complete image rebuild.

The RPM build uses fedora:latest without doing any kind of update, and until recently that was loading v31, there are only two docker hosts however, and in this PR we've done "docker build --pull fedora:latest", which will update the image on the docker host, and affect any job which is using fedora:latest, and that will include master.

**Comment by ashleypittman** [comment]
I've taken the --pull from the docker build command like as @JohnMalmberg is working on this seperately.

**Comment by brianjmurrell** [comment]
> The RPM build uses fedora:latest without doing any kind of update, and until recently that was loading v31, there are only two docker hosts however, and in this PR we've done "docker build --pull fedora:latest", which will update the image on the docker host, and affect any job which is using fedora:latest, and that will include master.

Was the `--pull` added on the stages that use `Dockerfile.mockbuild` also?  I thought it was only the `Dockerfile.checkpatch` stage.

Because if it was not, the docker image would have been rebuilt only if the docker image for that Dockerfile needed to be rebuilt for some reason.  TTBOMK nothing has changed in that Dockerfile that would have prompted an image rebuild.  And if an image was rebuilt since fedora:latest got updated to f33, then docker is not working the way we are expecting (as there should have been no change, as mentioned previously) and we should investigate why.

File: Jenkinsfile
Comment (brianjmurrell) [inline Jenkinsfile:564]: You still need the `--pull` to check our Nexus DockerHub proxy for updates to the base image don't you?  Otherwise we are back to never checking (or Nexus DockerHub proxy now) if the base docker image is newer aren't we?

All that proxying on Nexus does is shields DockerHub from all of our `--pull` hits.  Well, as well as theoretically improving rebuild performance by making the pull local.

What is unclear at this point is how we point our docker builds at our proxy instead of defaulting to DockerHub.  I guess we need to provide an argument of some sort pointing at Nexus.  I would think that needs to be here in this `additionalBuildArgs`.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

**Comment by ashleypittman** [comment]
> > The RPM build uses fedora:latest without doing any kind of update, and until recently that was loading v31, there are only two docker hosts however, and in this PR we've done "docker build --pull fedora:latest", which will update the image on the docker host, and affect any job which is using fedora:latest, and that will include master.
> 
> Was the `--pull` added on the stages that use `Dockerfile.mockbuild` also? I thought it was only the `Dockerfile.checkpatch` stage.
> 
> Because if it was not, the docker image would have been rebuilt only if the docker image for that Dockerfile needed to be rebuilt for some reason. TTBOMK nothing has changed in that Dockerfile that would have prompted an image rebuild. And if an image was rebuilt since fedora:latest got updated to f33, then docker is not working the way we are expecting (as there should have been no change, as mentioned previously) and we should investigate why.

I did a --pull in the checkpatch stage and that updated fedora:latest on one of the builders.  When building Docker.mockbuild it presumably checked the sha1 on the base image and re-built it as it was out of date.  That's exactly how I'd expect Docker to behave.

File: Jenkinsfile
Comment (ashleypittman) [reply Jenkinsfile:564]: As I said @JohnMalmberg is working on this.  The hope is we can have a weekly "docker --pull" job which will refresh everything once a week, and then all images will use the same base (updating the contents of course).  Whilst that may not be in place yet the correct thing for this PR is to assume that fedora:latest is correct and just use it.

Comment (brianjmurrell) [reply Jenkinsfile:564]: > As I said @JohnMalmberg is working on this. The hope is we can have a weekly "docker --pull" job which will refresh everything once a week, and then all images will use the same base (updating the contents of course). Whilst that may not be in place yet the correct thing for this PR is to assume that fedora:latest is correct and just use it.

FWIW, I don't think this is the correct approach.  Our docker registry is not a "mirroring" process but a "caching" process.  It doesn't _periodically_ refresh what it has, it refreshes what it has based on requests from clients and the age of what it's caching.  If an image goes out of use, it automatically stops getting refreshed as it's never asked for.

So currently, we have these (unnecessary) jobs running on the docker hosts, refreshing their caches, unnecessarily.  It is refreshing caches of a whilelist of images.  That's a whitelist that needs ongoing maintenance as we add more/different docker images.  That refresh is causing pulls to be done for things that might not necessarily be getting used all that often,  and I suspect that problem will only continue to grow as the whitelist will likely never be pruned of images that we stop using.

So ultimately with this method the rate of hits we make to dockerhub will continue to grow and grow (unbounded) as we artificially refresh both the docker host caches and the Nexus registry cache of images we are no longer interested in (because they are not actually in use other than the artificial cache refreshing).  One day this will become a **fire** that needs putting out because we will be refreshing so many unused images that we max out our Docker Hub quota pulling images no longer in use.

Contrast that to my proposal.  Instead of artificially refreshing caches, we let the components function as they are intended to.  That is, every `docker build` includes `--pull`.  That forces the docker hosts to see if they need to pull a new image from Nexus and pull it if necessary -- on every `docker build` attempt.  This keeps the docker host caches fresh automatically, without whitelists that need maintenance, including pruning.  All of those pulls will be satisfied by Nexus without hitting DockerHub for as long as Nexus' cache for that item is considered fresh.  That can be daily, weekly, whatever we want.

On every single one of those pulls (and only on pulls for jobs that actually want the image), when that expiry time for an item expires, Nexus will see if it needs to pull from Docker Hub to refresh it's own cache.  If it does, it passes that back through the docker hosts, refreshing their caches and to the client, to build a new image.

Basically Nexus becomes our rate-limiter for DockerHub while allowing us to hit is as many times as we want for "fresh" images.

In this way, the number of pulls we do from DockerHub is limited to only the number of images currently being used at a frequency of whatever their cache expiry time is in Nexus.  That number (currently_used_images x frequency) becomes the maximum number of hits we make to Docker Hub instead of the alternative above (number_of_images_in_unpruned_whitelist x frequency) where we are making a hit for everything that is in that (potentially, but probably, likely) unpruned whitelist at the same frequency.

Comment (kprantis) [reply Jenkinsfile:564]: I would suggest, if we are going to change how we do something at a higher level in how we manage images for reasons that exist the same both before and after this change, that we don't hold up existing PRs to have this debate and let them proceed, and such a debate/change should be its own isolated project.

File: Jenkinsfile
Comment (kprantis) [reply Jenkinsfile]: I will point out, though this conversation is a bit older now, there is also an in between I think, where it could update weekly, and we could just register for PSIRT notifications for the rare event there is something that can't wait a week, if it's going to cause challenges to our systems to handle keeping them that up to date (whether due to brittleness, or load, or complexity, or just in general making things take longer on a regular basis). These things may not be as an obvious a disruption to the team as a Jenkins update, but can be impactful all the same...I hope we are weighing those things too.

Comment (brianjmurrell) [reply Jenkinsfile]: > we could just register for PSIRT notifications for the rare event there is something that can't wait a week

I'm not sure what that means really, what granularity it's at, etc.  If it's something that needs action every time a new O/S and/or release/version is brought into our CI pipeline, that sounds like something that is going to fail to continue to get updated.  There is just too much of the left hand not knowing what the right hand is doing around here with many people updating Dockerfiles, CI test nodes, builds, etc.

On the other hand if it's a notification about vulnerabilities in every release of every O/S out there, that sounds like a haystack of notifications and somebody has to go through all of our uses of O/S releases to see what's vulnerable and what's not.

And then there is the case where we are updating say, weekly and a PSIRT does come out.  That becomes one of those very disruptive _everyone drop everything_ to fix it situations that chews up a few man-days.

These are all of the things I am trying to avoid.  Updating frequently, I don't think is unachievable.  We have lots of caching around here that should hopefully combat brittleness.  And we currently already update most of our images every time we use them (so as frequently as is actually possible) and that seems to be working fine, other than the update time gets longish when base images are not updated frequently enough.  But that's a solvable problem, with resources to apply to it.

**Review by brianjmurrell — APPROVED** [review]

File: Jenkinsfile
Comment (brianjmurrell) [reply Jenkinsfile:564]: I'm not sure I entirely parsed all of that correctly, but I didn't think we were "holding anything up" here.  This came across as an general enhancement that didn't have any kind of deadline to landing.  The ticket for this says it's targeted for 2.0.

But certainly, if I am mistaken about that, this indeed could land as is, as it ticks the boxes of ensuring the image is always up-to-date as it is, and the refinement of how we fetch images and keep them up-to-date can happen after landing this.

Comment (kprantis) [reply Jenkinsfile:564]: It's not urgent in the sense of a release requirement, but, I think, has a lot of benefit we could really use to start benefiting the team immediately, and then perhaps work on such things as follow-ons, if we want to do them.

File: Jenkinsfile
Comment (kprantis) [reply Jenkinsfile]: I could be wrong, but, I thought it had some granularity to what projects/products you could subscribe to. It certainly would require someone to own following up on them, but since we already need to get those for our dependencies to get notified of CVEs, I would think we could just roll that in to the same ownership/checking. And I don't think that having to update suddenly when it's updating weekly already is likely to be a very onerous task, no? Not like there's likely to be lots of regressions to yak-shave, etc. So while unexpected, I would think they'd be more likely to be a pretty simple activity, no?

And not necessarily saying we shouldn't automate these updates, just want to make sure we're also weighing those pros/cons with other options. Longish updates are one of those cons that are rather impactful. It is solvable, yes, but not solved yet. So perhaps there is an ordering of prioritizing these improvements that can minimize impact to our own development...

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**
